### PR TITLE
Enable sorting and filtering for various kind of tests

### PR DESCRIPTION
**Pull request by raoyitao**
* Update sorting and filter module so that they can deal with basic
  string type as well as testsuite/testcase objects. Related code
  is refactored to support any custom type of testsuite/testcase.
* Test sorter can be applied to PyTest and PyUnit.
* Test filter can be applied to PyTest and PyUnit (but PyUnit cannot
  support testcase-level filtering).
* A new decorator `dispatchmethod` is added to help an instance method
  to deal with arguments of different, it might be removed after our
  supported interpreter is updated to 3.9 or 3.10.
* Remove a fixture that is no longer used.
* Add and update testcases.

## Checklist:
- [x] Test
- [ ] Example (both test_plan.py and .rst)
- [ ] Documentation (API)
- [ ] News fragment present for release notes
- [ ] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [ ] For new cmdline arg: documentation

### CODE REVIEW

File: tests/unit/testplan/testing/test_pytest.py [hunk 1]
```diff
@@ -5,42 +5,45 @@
...
 
+PYTEST_DEFAULT_PARAMS = {
+    "test_filter": filtering.Filter(),
+    "test_sorter": ordering.NoopSorter(),
+    "stdout_style": defaults.STDOUT_STYLE,
+}
+
 
 @pytest.fixture
-def pytest_test_inst(repo_root_path, root_directory):
```
Comment (Pyifan) [inline tests/unit/testplan/testing/test_pytest.py:15]: I remember root_directory is needed for internal test.

File: tests/unit/testplan/testing/test_pytest.py [hunk 2]
```diff
@@ -96,7 +102,7 @@ def test_run_testcases_iter_testcase(pytest_test_inst):
     """Test running a single testcase iteratively."""
     all_results = list(
         pytest_test_inst.run_testcases_iter(
-            testsuite_pattern="pytest_tests.py::TestPytestBasics",
+            testsuite_pattern="examples/PyTest/pytest_tests.py::TestPytestBasics",
```
Comment (Pyifan) [inline tests/unit/testplan/testing/test_pytest.py:105]: What is the impact of this change? Can you show me the test_plan_top.py report please?

**Review by Pyifan — APPROVED** [review]

File: tests/unit/testplan/testing/test_pytest.py (hunk 2 above)
Comment (raoyitao) [reply tests/unit/testplan/testing/test_pytest.py:107]: http://testplan-db.ms.com/testplan/15bf8a587ff549faaf321f28a3b3c2e9
here is an example

File: tests/unit/testplan/testing/test_pytest.py (hunk 1 above)
Comment (raoyitao) [reply tests/unit/testplan/testing/test_pytest.py:15]: yes. this is restored.

File: tests/unit/testplan/testing/test_pytest.py (hunk 2 above)
Comment (Pyifan) [reply tests/unit/testplan/testing/test_pytest.py:107]: the pytest name and testsuite name duplicates, is there any way to refine this?

**Review by Pyifan — APPROVED** [review]

File: tests/unit/testplan/testing/test_pytest.py (hunk 2 above)
Comment (Pyifan) [reply tests/unit/testplan/testing/test_pytest.py:107]: The new fragment needs to be renamed to conform with latest convension.

File: doc/newsfragments/fix_debug_log_in_interactive_mode.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+* Fix an issue that in interactive mode, debug log is unexpectedly printed even without "-d" switch.
```
Comment (Pyifan) [inline doc/newsfragments/fix_debug_log_in_interactive_mode.rst:1]: the news fragment need to conform with new convention.

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: tests/unit/testplan/testing/test_pytest.py (hunk 2 above)
Comment (raoyitao) [reply tests/unit/testplan/testing/test_pytest.py:107]: fragment updated. the pytest name is the start string of test suite name. for now I don't have a good idea for their names.

**Review by Pyifan — APPROVED** [review]

**Pull request merged**